goldmane/pkg/aggregator/index_test.go

```diff
@@ -160,11 +185,404 @@ func TestIndexAddRemove(t *testing.T) {
 	// Verify that the index is empty.
-	flows = idx.List(IndexFindOpts{})
+	flows, _ = idx.List(IndexFindOpts{})
 	Expect(flows).To(HaveLen(0))
 
 	// Remove flows we know aren't in the index, to make sure we're idempotent.
 	for _, flow := range allFlows {
 		idx.Remove(flow)
 	}
 }
+
+func TestIndexPagination_General(t *testing.T) {
+	defer setupTest(t)()
+
+	allFlows := []*types.DiachronicFlow{
+		{
+			ID:  0,
+			Key: types.FlowKey{DestName: "a", DestNamespace: "ns1"},
+		},
+		{
+			ID:  1,
+			Key: types.FlowKey{DestName: "c", DestNamespace: "ns1"},
+		},
+		{
+			ID:  2,
+			Key: types.FlowKey{DestName: "b", DestNamespace: "ns1"},
+		},
+		{
+			ID:  3,
+			Key: types.FlowKey{DestName: "d", DestNamespace: "ns1"},
+		},
+		{
+			ID:  4,
+			Key: types.FlowKey{DestName: "e", DestNamespace: "ns2"},
+		},
+		{
+			ID:  5,
+			Key: types.FlowKey{DestName: "f", DestNamespace: "ns2"},
+		},
+		{
+			ID:  6,
+			Key: types.FlowKey{DestName: "g", DestNamespace: "ns2"},
+		},
+		{
+			ID:  7,
+			Key: types.FlowKey{DestName: "h", DestNamespace: "ns2"},
+		},
+		{
+			ID:  8,
+			Key: types.FlowKey{DestName: "i", DestNamespace: "ns3"},
+		},
+	}
+
+	data := &types.Flow{
+		PacketsIn:               1,
+		PacketsOut:              1,
+		BytesIn:                 1,
+		BytesOut:                1,
+		NumConnectionsLive:      1,
+		NumConnectionsStarted:   1,
+		NumConnectionsCompleted: 1,
+	}
+
+	// Create an index, ordered by destination name.
+	idx := NewIndex(func(k *types.FlowKey) string {
+		return k.DestName
+	})
+
+	for _, flow := range allFlows {
+		// Make sure each one has some data in it.
+		flow.AddFlow(data, 0, 1)
+		idx.Add(flow)
+	}
+
+	tt := []struct {
+		description          string
+		page                 int64
+		pageSize             int64
+		expectedTotalPages   int
+		expectedTotalResults int
+		expectedNumFlows     int
+		filter               *proto.Filter
+	}{
+		{
+			description:          "page 0, limit 1",
+			page:                 0,
+			pageSize:             1,
+			expectedTotalPages:   9,
+			expectedTotalResults: 9,
+			expectedNumFlows:     1,
+		},
+		{
+			description:          "page 0, limit 2",
+			page:                 0,
+			pageSize:             2,
+			expectedTotalPages:   5,
+			expectedTotalResults: 9,
+			expectedNumFlows:     2,
+		},
+		{
+			description:          "page 0, limit 2",
+			page:                 0,
+			pageSize:             2,
+			expectedTotalPages:   2,
+			expectedTotalResults: 4,
+			expectedNumFlows:     2,
+			filter:               &proto.Filter{DestNamespaces: []*proto.StringMatch{{Value: "ns2"}}},
+		},
+		{
+			description:          "page 0, size 0",
+			page:                 0,
+			pageSize:             0,
+			expectedTotalPages:   1,
+			expectedTotalResults: 4,
+			expectedNumFlows:     4,
+			filter:               &proto.Filter{DestNamespaces: []*proto.StringMatch{{Value: "ns2"}}},
+		},
+	}
+
+	for _, tc := range tt {
+		t.Run(tc.description, func(t *testing.T) {
+			RegisterTestingT(t)
```

Comment:
I don't think we need this - it's in setupTest?

Reply:
Right, but we should be calling setupTest() within this t.Run anyway

Reply:
Ya I realized that looking into your other comment.